src/runtime/composables/useSupabaseUser.ts

```diff
@@ -1,9 +1,28 @@
 import type { User } from '@supabase/supabase-js'
 import { useSupabaseSession } from './useSupabaseSession'
-import { computed, type ComputedRef } from '#imports'
+import { useSupabaseClient } from './useSupabaseClient'
+import { useState } from '#imports'
+
+export const useSupabaseUser = async () => {
+  const userState = useState<User | null>('supabase_user', () => null)
+  const supabase = useSupabaseClient()
+
+  // Need to manipulate user in `supabase.client` plugin when client is not yet initialized
+  if (!supabase) {
+    return userState
+  }
+
+  // We do not rely on `getSession` on the server. It could be tampered with by the sender.
+  // User has been populated by `getUser` in server plugin.
+  if (import.meta.server) {
+    return userState
+  }
+  // We rely on `getSession` on the client.
+  else {
+    const session = await useSupabaseSession()
```

Comment:
If we already get the user info for the server in the plugin, does it make sense to also do the same in the frontend? This would clean up this composable a bit and I am not sure if there is value in refreshing the userState each time useSupabaseUser is called. It might even be bad for performance.

Reply:
From my point of view, `onAuthStateChange` is only triggered when Supabase API is called to refresh the session either when `getSession` or `refreshSession` are called or when user is logged out. But on client side, when navigating, we want to ensure that the session (and the user) are still connected (token not expired) that's why calling `getSession` ensures this. 

Important to note that [`getSession`](https://supabase.com/docs/reference/javascript/auth-getsession) is not calling Supabase API when token are not expired (most of the time) but only reading local storage, this is not a lack of performance thus. 

However, if you want to make some tests and prove that `onAuthStateChange` is triggered when token is expired and not refresh, we could remove it.

Reply:
From what I can tell, the [`onAuthStateChange`](https://supabase.com/docs/reference/javascript/auth-onauthstatechange) is called on these events that would cover token expiration because `SIGNED_OUT` is also called when the token expires. Reading the docs, it definitely looks to me like one can rely on these auth events to track sessions. We also might not even need to call a `getSession` to initialize since the `INITIAL_SESSION` auth event triggers after client creation.

I can do more testing later today if you'd like

Reply:
In this case we can indeed remove the async call to `useSupabaseSession` and get `useSupabaseUser` back to a synchronous composable. Which is a good thing meaning no breaking changes!

You can update the code then we'll both update our personal projects to ensure it works as expected.

Reply:
`useSupabaseUser` will be back to the old version only returning the state. State being muted by either the server plugin or the `onAuthStateChange`. It makes sense to me. `useSupabaseSession` stays as a standalone composable.

Reply:
I was thinking the same writing my previous message but it seems ok to me to have a 100% accurate session from a Supabase client call in this case. Indeed we're only using the `useSupabaseSession` in the `auth-redirect` middleware, at least we ensure we are up to date at that moment (no delay from the hook). Wdyt?

Reply:
I made the changes on a separate branch (https://github.com/felixgabler/supabase-nuxt/commit/0895cccd0573aafa0551a4c2496d23cffc7f0717). Will test it out in my local project because I prefer the general approach and cleaner code :)

Reply:
@felixgabler If you confirm everything works smoothly with your project, let's push your changes in this PR. It looks all good to me! I'll test it on mine once you push your changes.

Reply:
We've been running it for the last few days without issues related to the changes. We do get the supabase `useSession` warnings which are annoying but not a problem. Looking forward to hearing how it works on your project :)

Reply:
I've pushed the new version on my side project. I'll tell you if everything's ok :) 

This warning is really annoying, they do not seem to be in a hurry with it...

Reply:
Hi! It would be awesome if we could move this forward. I don't think they will remove the warning as it is generally a bit unsafe what supabase-nuxt does.

The warning only shows on server side if `session.user` is accessed (because this is unsafe behavior). While this PR does not actually read the user from the session, we do write the session using `useState` (see [this line](https://github.com/nuxt-modules/supabase/pull/357/files#diff-3ce79a8bc112e95889382b506a02bb39d84601b9cc48fd40f2d243fb8be06bcbR38)). Any serialization of the session would also cause it to evaluate its `user` field, therefore triggering the warning.

I propose to make a breaking change and making useSession only work client side. I don't really see any other answer because Supabase (rightly, I believe) will not remove this warning and at the same time, the latest changes to `@supabase/ssr` are super important and full of bug fixes. My project runs much smoother since we switched.

What do you think?

Reply:
I missed this one @felixgabler sorry. Then we're aligned (see my last message on the main thread). Let's merge it. My opinion is not to create breaking changes and just release it with the warning. People will complain about it and it might add upvotes to our issue so that supabase maintainers will address it!